name: branch
on:
  workflow_call:
    secrets:
      NPM_TOKEN:
        required: true
      GIT_TOKEN:
        required: true
    inputs:
      cleanCheckout:
        description: "Whether to prefer a clean checkout (debug purposes)"
        default: false
        required: false
        type: boolean
      qaStorybook:
        description: "Whether to build and deploy a QA Storybook doc"
        default: false
        required: false
        type: boolean
env:
  NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
  GITHUB_TOKEN: ${{ secrets.GIT_TOKEN }}
jobs:
  test:
    runs-on: [self-hosted, frontend]
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          clean: ${{inputs.cleanCheckout}}
          token: ${{secrets.GIT_TOKEN}}
      - name: Node
        uses: actions/setup-node@v2
        with:
          node-version: "16"
          registry-url: "https://registry.npmjs.org"
        env:
          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
      - name: Install
        id: install
        continue-on-error: true
        run: npm install --legacy-peer-deps
      - name: Install-Retry1
        continue-on-error: true
        id: installRetry1
        if: steps.install.outcome=='failure'
        run: npm rebuild && npm cache clean --force && npm i --legacy-peer-deps
      - name: Test
        id: test
        continue-on-error: true
        run: |
          git fetch origin main:main
          npm run test
      - name: Test-Retry1
        continue-on-error: true
        id: testRetry1
        if: steps.test.outcome=='failure'
        run: npm run test
      # If you got to this point, we'll do a clean checkout to clear the git cache.
      # This legitimately has fixed multiple build-time sanity errors.
      - name: DesperateCheckout
        if: steps.testRetry1.outcome=='failure'
        uses: actions/checkout@v2
        with:
          clean: true
          token: ${{secrets.GIT_TOKEN}}
      - name: Test-Retry2
        id: testRetry2
        if: steps.testRetry1.outcome=='failure'
        run: npm run test
      - name: Coverage
        continue-on-error: true
        id: coverage
        uses: raulanatol/jest-coverage-comment-action@main
        with:
          github-token: ${{secrets.GIT_TOKEN}}
          jest-command: "npm run test:unit:coverage"
      - name: QA-Storybook
        if: ${{ inputs.qaStorybook }}
        continue-on-error: true
        run: |
          npm run build-storybook
          usafacts azsync --source "${GITHUB_WORKSPACE}/storybook-static" --container "storybook/qa/${GITHUB_REPOSITORY#*/}"
